### PR TITLE
mysql: MyDecimal implements multiply, division and modulus.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @tiancaiamao @BusyJay PTAL

File: mysql/mydecimal.go [hunk 1]
```diff
@@ -1086,7 +1100,7 @@ func (d *MyDecimal) ToBin(precision, frac int) ([]byte, int) {
 	leadingDigitsFrom := digitsIntFrom - wordsIntFrom*digitsPerWord
 	iSizeFrom := wordsIntFrom*wordSize + dig2bytes[leadingDigitsFrom]
 
-	if wordsInt < wordsIntFrom {
+	if digitsInt < digitsIntFrom {
```
Comment (shenli) [inline mysql/mydecimal.go]: Any  test case cover this？

File: mysql/mydecimal.go [hunk 2]
```diff
@@ -1634,3 +1648,425 @@ func maxDecimal(precision, frac int, to *MyDecimal) {
...
+		wordsInt1   = (from1.digitsInt + digitsPerWord - 1) / digitsPerWord
+		wordsFrac1  = (from1.digitsFrac + digitsPerWord - 1) / digitsPerWord
+		wordsInt2   = (from2.digitsInt + digitsPerWord - 1) / digitsPerWord
+		wordsFrac2  = (from2.digitsFrac + digitsPerWord - 1) / digitsPerWord
+		wordsIntTo  = (from1.digitsInt + from2.digitsInt + digitsPerWord - 1) / digitsPerWord
+		wordsFracTo = wordsFrac1 + wordsFrac2
+		idx1        = wordsInt1
+		idx2        = wordsInt2
+		idxTo       = 0
+		iii         = wordsIntTo
```
Comment (shenli) [inline mysql/mydecimal.go]: Weird name.

**Comment by coocood** [comment]
@shenli PTAL

File: mysql/mydecimal.go [hunk 3]
```diff
@@ -96,6 +97,23 @@ func sub(a, b, carry int32) (int32, int32) {
 	return diff, carry
 }
 
+// sub2 substracts b and carry from a, returns the diff and new carry.
```
Comment (zimulala) [inline mysql/mydecimal.go]: carry -> carries ?

File: mysql/mydecimal.go [hunk 4]
```diff
@@ -1634,3 +1648,425 @@ func maxDecimal(precision, frac int, to *MyDecimal) {
...
+    E_DEC_OK/E_DEC_TRUNCATED/E_DEC_OVERFLOW;
+
+  NOTES
+    in this implementation, with wordSize=4 we have digitsPerWord=9,
+    and 63-digit number will take only 7 words (basically a 7-digit
+    "base 999999999" number).  Thus there's no need in fast multiplication
+    algorithms, 7-digit numbers can be multiplied with a naive O(n*n)
+    method.
+
+    XXX if this library is to be used with huge numbers of thousands of
```
Comment (tiancaiamao) [inline mysql/mydecimal.go]: XXX?

File: mysql/mydecimal_test.go [hunk 5]
```diff
@@ -554,3 +558,87 @@ func (s *testMyDecimalSuite) TestSub(c *C) {
...
+	type tcase struct {
+		a      string
+		b      string
+		result string
+		ec     int
+	}
+	cases := []tcase{
+		{"12", "10", "120", 0},
+		{"-123.456", "98765.4321", "-12193185.1853376", 0},
+		{"-123456000000", "98765432100000", "-12193185185337600000000000", 0},
```
Comment (tiancaiamao) [inline mysql/mydecimal_test.go]: is there any test case that covers overflow?  which is more than 81 digits a decimal can represent

**Comment by tiancaiamao** [comment]
LGTM

File: mysql/mydecimal_test.go [hunk 6]
```diff
@@ -15,6 +15,7 @@ package mysql
 
 import (
 	. "github.com/pingcap/check"
+	"strings"
```
Comment (shenli) [inline mysql/mydecimal_test.go]: Put strings before "github.com/pingcap/check"

File: mysql/mydecimal.go [hunk 7]
```diff
@@ -1634,3 +1648,427 @@ func maxDecimal(precision, frac int, to *MyDecimal) {
 		}
 	}
 }
+
+/*
+DecimalMul multiply two decimals.
```
Comment (shenli) [inline mysql/mydecimal.go]: multiplies

File: mysql/mydecimal.go [hunk 8]
```diff
@@ -1634,3 +1648,427 @@ func maxDecimal(precision, frac int, to *MyDecimal) {
...
+			to.wordBuf[curIdx] = to.wordBuf[idxTo]
+			curIdx++
+			idxTo++
+			dToMove--
+		}
+	}
+	return errCode
+}
+
+// DecimalDiv do division of two decimals.
```
Comment (shenli) [inline mysql/mydecimal.go]: does

**Comment by shenli** [comment]
LGTM

**Pull request merged**